### PR TITLE
DDF-1313: Added an OSGi command to create a local GeoNames index

### PR DESCRIPTION
**Pull request by jrnorth**
@pklinef @stustison @millerw8 @jlcsmith @kcwire @wmcnalli @beyelerb @emanns95 @kguerriero

[Review on Reviewable](https://reviewable.io/reviews/codice/ddf-catalog/118)

### CODE REVIEW

File: spatial/geo/spatial-geo-index/pom.xml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.lucene</groupId>
+            <artifactId>lucene-core</artifactId>
+            <version>5.2.1</version>
```
Comment (jrnorth) [inline spatial/geo/spatial-geo-index/pom.xml]: There's a subtle issue that's been introduced by upgrading the Lucene version. If this bundle is running using Lucene 5.2.1 and it gets reloaded via a `dev:watch`, the next time you try to use the `update` command there is a `ClassCastException` somewhere in the Lucene code. I think it has to do with how I'm packaging the Lucene jars in the bundle; I'll look into it.

**Comment by pklinef** [comment]
spatial-geo-* seems redundant and unclear.  I would suggest something like spatial-geocoding-* instead.

**Comment by pklinef** [comment]
Where are the unit tests and documentation?

**Comment by jrnorth** [comment]
Where would be the most appropriate place to put documentation for something like this?

File: spatial/pom.xml [hunk 2]
```diff
@@ -158,5 +158,7 @@
         <module>csw</module>
         <module>spatial-app</module>
         <module>docs</module>
```
Comment (pklinef) [inline spatial/pom.xml]: Documentation for administrators can be found in this module.

**Comment by jrnorth** [comment]
Working on unit tests, documentation, and renaming my modules now.

---

Review status: 0 of 25 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[spatial/geo/spatial-geo-index/pom.xml, line 43 \[r1\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JuIgCUvScJAkeTCRmi1-r1-43)** ([raw file](https://github.com/codice/ddf-catalog/blob/4d379d4054810f64ae6322e4529fd452d235401b/spatial/geo/spatial-geo-index/pom.xml#L43)):</sup>
This has been fixed by https://github.com/codice/ddf-catalog/commit/48e17f6946889424774cae5bd9fc4d3b64f28282.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by kcwire** [comment]
Reviewed 21 of 24 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

<sup>**[spatial/geo/spatial-geo-api/pom.xml, line 57 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug4zUaMsWA3pPYlSq2)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-api/pom.xml#L57)):</sup>
The api is probably the only place these limits should be allowed, otherwise they should be removed.

---

<sup>**[spatial/geo/spatial-geo-create/pom.xml, line 62 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug5WSSeMzaj6wdSv9J)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-create/pom.xml#L62)):</sup>
remove these.

---

<sup>**[spatial/geo/spatial-geo-extract/pom.xml, line 79 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug5jYZlQrlRoPvThmP)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-extract/pom.xml#L79)):</sup>
remove

---

<sup>**[spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java, line 84 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug6E5nTVr-9emcxLNj)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java#L84)):</sup>
is there any issue with loading a very large file into the lineNumberReader?

---

<sup>**[spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java, line 110 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug6cXIeEyCJRbA305T)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java#L110)):</sup>
You could probably group these exceptions together. I would expect the exception info to be informative enough that you could use a generic log message and include the exception.

---

<sup>**[spatial/geo/spatial-geo-index/pom.xml, line 42 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug79lQN8IiPaw7rfwo)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-index/pom.xml#L42)):</sup>
@pklinefelter - shouldn't we just use the Lucene dependecies directly and embed it vs trying to import it from another bundle?

---

<sup>**[spatial/geo/spatial-geo-index/pom.xml, line 81 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug7NqgOk3Qysl2hBC6)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-index/pom.xml#L81)):</sup>
you will write unit tests and remove these :)

---

<sup>**[spatial/geo/spatial-geo-index/src/main/java/org/codice/ddf/spatial/geo/index/GeoNamesLuceneIndexer.java, line 69 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug7Xi6m3h7_BMYilLo)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-index/src/main/java/org/codice/ddf/spatial/geo/index/GeoNamesLuceneIndexer.java#L69)):</sup>
You seem to be doing a lot of logging and rethrowing.  Typically if you log then you wouldn't throw, or if you need to throw then you shouldn't log, since the calling application should be logging it for you.  Double check your log statements and make sure we are only logging these exceptions once in the call stack.

---

<sup>**[spatial/spatial-commands/pom.xml, line 77 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug8VDsQReVS0Kj1ms8)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/spatial-commands/pom.xml#L77)):</sup>
remove

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by jrnorth** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

<sup>**[spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java, line 84 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug6E5nTVr-9emcxLNj)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-extract/src/main/java/org/codice/ddf/spatial/geo/extract/GeoNamesFileExtractor.java#L84)):</sup>
No, I haven't encountered any issues.

---

<sup>**[spatial/geo/spatial-geo-index/pom.xml, line 42 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jug79lQN8IiPaw7rfwo)** ([raw file](https://github.com/codice/ddf-catalog/blob/48e17f6946889424774cae5bd9fc4d3b64f28282/spatial/geo/spatial-geo-index/pom.xml#L42)):</sup>
Embedding lucene-core and lucene-analyzers-common is what I was doing previously, but I ran into issues with Lucene not being able to find classes it needed if the bundle was reloaded.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by wmcnalli** [comment]
Reviewed 20 of 44 files at r3.
Review status: 21 of 26 files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by wmcnalli** [comment]
:+1: 

---

Review status: 21 of 26 files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by kcwire** [comment]
Reviewed 3 of 44 files at r3, 12 of 18 files at r4.
Review status: 30 of 36 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[spatial/geocoding/spatial-geocoding-create/src/test/java/org/codice/ddf/spatial/geocoding/create/TestGeoNamesCreator.java, line 40 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JurwRKMPR7BHTbBzgQn)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-create/src/test/java/org/codice/ddf/spatial/geocoding/create/TestGeoNamesCreator.java#L40)):</sup>
what about some negative test cases? wrong format? empty line? etc

---

<sup>**[spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java, line 93 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-Jurx7DWBkKleQ_h3yiE)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java#L93)):</sup>
instead of constructing the paths like this you could do `URL url = Thread.currentThread().getContextClassLoader().getResource("mypackage/YourFile.csv");
url.getPath();` in an @BeforeClass method to ensure all of your resources exist.

---

<sup>**[spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java, line 145 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JurxfE0RDmEhrbFm6GV)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java#L145)):</sup>
It's generally bad practice to assert on the Exception message.  If the type of exception isn't descriptive enough then we may need to create our own.

---

<sup>**[spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java, line 149 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JurxqygIkuZh8Q2n7sN)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-extract/src/test/java/org/codice/ddf/spatial/geocoding/extract/TestGeoNamesFileExtractor.java#L149)):</sup>
This is much better.

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexer.java, line 110 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JuryT_RW7Ka30pZVt-C)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexer.java#L110)):</sup>
you are asserting on the same strings & doubles a lot. Make them constants so when you have to change `Phoenix` to `Feenix` you only have to change it once.

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexer.java, line 122 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JurylZnVXOOxPnOaHTg)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexer.java#L122)):</sup>
3 seems magical. Assert on the list.size()

---

<sup>**[spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexerExceptions.java, line 140 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/118#-JurzEexdLZf-dKdiuzW)** ([raw file](https://github.com/codice/ddf-catalog/blob/3a77ac11c69d0f5234b0a00408b37c712527e5c5/spatial/geocoding/spatial-geocoding-index/src/test/java/org/codice/ddf/spatial/geocoding/index/TestGeoNamesLuceneIndexerExceptions.java#L140)):</sup>
no assertions?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/118)

**Comment by kcwire** [comment]
if there are no objections I will merge this today.

**Comment by kcwire** [comment]
@jrnorth please resolve the merge conflicts.

**Pull request merged**